src/runtime/webgpu.cpp

```diff
@@ -0,0 +1,1146 @@
+        return error_code;
+    }
+
+    if (staging_buffer == nullptr) {
+        ErrorScope error_scope(user_context, context.device);
+
+        // Create a staging buffer for transfers if we haven't already.
+        WGPUBufferDescriptor desc{};
+        desc.nextInChain = nullptr;
+        desc.label = nullptr;
+        desc.usage = WGPUBufferUsage_CopyDst | WGPUBufferUsage_MapRead;
+        desc.size = kWebGpuStagingBufferSize;
+        desc.mappedAtCreation = false;
+
+        staging_buffer = wgpuDeviceCreateBuffer(global_device, &desc);
+
+        int error_code = error_scope.wait();
+        if (error_code != halide_error_code_success) {
+            staging_buffer = nullptr;
+            return error_code;
+        }
+    }
+
+    buf->device = (uint64_t)device_handle;
+    buf->device_interface = &webgpu_device_interface;
+    buf->device_interface->impl->use_module();
+
+    debug(user_context)
+        << "      Allocated device buffer " << (void *)buf->device << "\n";
+
+    return halide_error_code_success;
+}
+
+WEAK int halide_webgpu_device_free(void *user_context, halide_buffer_t *buf) {
+    if (buf->device == 0) {
+        return 0;
+    }
+
+    WgpuBufferHandle *handle = (WgpuBufferHandle *)buf->device;
+
+    debug(user_context)
+        << "WGPU: halide_webgpu_device_free (user_context: " << user_context
+        << ", buf: " << buf << ") WGPUBuffer: " << handle->buffer << "\n";
+
+    WgpuContext context(user_context);
+    if (context.error_code) {
+        return context.error_code;
+    }
+
+    wgpuBufferRelease(handle->buffer);
+    free(handle);
+    buf->device = 0;
+    buf->device_interface->impl->release_module();
+    buf->device_interface = nullptr;
+
+    return halide_error_code_success;
+}
+
+WEAK int halide_webgpu_device_sync(void *user_context, halide_buffer_t *) {
+    WgpuContext context(user_context);
+    if (context.error_code) {
+        return context.error_code;
+    }
+
+    ErrorScope error_scope(user_context, context.device);
+
+    // Wait for all work on the queue to finish.
+    struct WorkDoneResult {
+        volatile ScopedSpinLock::AtomicFlag complete = false;
+        volatile WGPUQueueWorkDoneStatus status;
+    };
+    WorkDoneResult result;
+
+    __atomic_test_and_set(&result.complete, __ATOMIC_RELAXED);
+    wgpuQueueOnSubmittedWorkDone(
+        context.queue, 0,
+        [](WGPUQueueWorkDoneStatus status, void *userdata) {
+            WorkDoneResult *result = (WorkDoneResult *)userdata;
+            result->status = status;
+            __atomic_clear(&result->complete, __ATOMIC_RELEASE);
+        },
+        &result);
+
+    int error_code = error_scope.wait();
+    if (error_code != halide_error_code_success) {
+        return error_code;
+    }
+
+    while (__atomic_test_and_set(&result.complete, __ATOMIC_ACQUIRE)) {
+        wgpuDeviceTick(context.device);
+    }
+
+    return result.status == WGPUQueueWorkDoneStatus_Success ?
```

Comment:
You should also call `halide_error()` or `error()` if the result isn't `halide_error_code_success` (yes, this is regrettably hard to get right -- see https://github.com/halide/Halide/issues/7350)